Handle licenses from git-deps for our use of next.jdbc

dpsutton opened the pull request:
```clojure
licenses=> (process* {:libs (-> basis :libs (select-keys '[com.github.seancorfield/next.jdbc]))})
{:with-license
 [[com.github.seancorfield/next.jdbc
   {:coords {:group "com.github.seancorfield", :artifact "next.jdbc"},
    :license
    "THE ACCOMPANYING PROGRAM IS PROVIDED UNDER THE TERMS OF THIS ECLIPSE PUBLIC..."}]],
 :without-license nil}
```

Previously it just filtered out non-jar deps.

some cleanup to make it easier to test, organizationally, etc.

Organizationally, now there is

```clojure
(defmulti lib-license
  "Determine the license from an lib. Implementations for both :mvn (jar) and :deps (git deps)."
  (fn [[_lib-name info]]
    (:deps/manifest info)))

(defmethod lib-license :mvn ...)
(defmethod lib-license :deps ...)
```

A few nicer cleanups. Searching through a jar filesystem and a git directory are both behind 
```clojure
(defprotocol FileSearch
  (-file-search [_ files]
    "Find the first file matching the names passed in. Returns a Path."))
```


And while i was in here, for each lib in our overrides.edn, I threw it out of the overrides and checked if we still got any warning.

#### Everything still works:

```
❯ clojure -X:build build/build! :edition :ee :steps '[:licenses]'
Warning: environ value /Users/dan/.sdkman/candidates/java/current for key :java-home has been overwritten with /Users/dan/.sdkman/candidates/java/17.0.1-zulu/zulu-17.jdk/Contents/Home
$ "git" "symbolic-ref" "--short" "HEAD"
  cleanup-license-check
$ "git" "describe" "--abbrev=0" "--tags"
  v0.45.0-RC2
Running build steps for Enterprise Edition version v1.45.1-SNAPSHOT: licenses
  Generate backend license information from jar files
    License information generated at /Users/dan/projects/work/metabase/resources/license-backend-third-party.txt
  Run `yarn licenses generate-disclaimer`
    $ "yarn" "licenses" "generate-disclaimer"
  All build steps finished.
```

```
❯ clojure -X:build build/list-without-license
All dependencies have licenses
```

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29343)
<!-- Reviewable:end -->

dpsutton committed:
Handle licenses from git-deps for our use of next.jdbc

```clojure
licenses=> (process* {:libs (-> basis :libs (select-keys '[com.github.seancorfield/next.jdbc]))})
{:with-license
 [[com.github.seancorfield/next.jdbc
   {:coords {:group "com.github.seancorfield", :artifact "next.jdbc"},
    :license
    "THE ACCOMPANYING PROGRAM IS PROVIDED UNDER THE TERMS OF THIS ECLIPSE PUBLIC..."}]],
 :without-license nil}
```

Previously it just filtered out non-jar deps.

some cleanup to make it easier to test, organizationally, etc.

dpsutton committed:
Add driver licenses and update calling code

both of the following need to work:
```
clojure -X:build build/list-without-license
clojure -X:build build/build! :edition :ee :steps '[:licenses]'
```

qnkhuat commented:
this looks stale, I don't have enough knowledge about licensing so I'm not sure I can fully review this. do you want to bump this to core-backend instead?

tsmacdonald commented:
@dpsutton what's the status of this? Happy to review if it's still relevant and you fix the merge conflicts

dpsutton committed:
Remove two unneeded overrides

```clojure
(ns nocommit.license-checks
  (:require [clojure.edn :as edn]
            [clojure.java.io :as io]
            [clojure.tools.build.api :as b]))

license-checks=> (def basis (b/create-basis {:project "deps.edn"
                                             :aliases ["ee" "drivers"]}))
license-checks=> (def libs (lic/process*
                             {:libs     (:libs basis)
                              :backfill (edn/read-string
                                          (slurp (io/resource "overrides.edn")))}))
license-checks=> (->> libs :without-license (map first))
()
```

dpsutton committed:
include ee and drivers profiles in checking for deps without license

dpsutton commented:
@tsmacdonald i've fixed the conflicts. Should be good to review now.

Removed some of the added libs as they are not needed.

camsaul reviewed (approved):
This is cool, definitely a nice lil improvement
